etcdutl/etcdutl/common_test.go

```diff
@@ -141,3 +151,184 @@ func TestGetLatestWalSnap(t *testing.T) {
 		})
 	}
 }
+
+func TestCreateV2SnapshotFromV3Store(t *testing.T) {
+	testCases := []struct {
+		name            string
+		consistentIndex uint64
+		term            uint64
+		clusterVersion  string
+		members         []uint64
+		learners        []uint64
+		removedMembers  []uint64
+		expectedErrMsg  string
+	}{
+		{
+			name:            "unexpected term: less than the last snapshot.term",
+			consistentIndex: 3,
+			term:            1,
+			expectedErrMsg:  "less than the latest snapshot",
+		},
+		{
+			name:            "unexpected consistent index: less than the last snapshot.index",
+			consistentIndex: 1,
+			term:            3,
+			expectedErrMsg:  "less than the latest snapshot",
+		},
+		{
+			name:            "normal case",
+			consistentIndex: 32,
+			term:            4,
+			clusterVersion:  "3.5.0",
+			members:         []uint64{100, 200},
+			learners:        []uint64{300},
+			removedMembers:  []uint64{400, 500},
+		},
+		{
+			name:            "empty cluster version",
+			consistentIndex: 45,
+			term:            4,
+			clusterVersion:  "",
+			members:         []uint64{110, 200},
+			learners:        []uint64{350},
+			removedMembers:  []uint64{450, 500},
+		},
+		{
+			name:            "no learner",
+			consistentIndex: 7,
+			term:            5,
+			clusterVersion:  "3.5.0",
+			members:         []uint64{150, 200},
+			removedMembers:  []uint64{450, 550},
+		},
+		{
+			name:            "no removed members",
+			consistentIndex: 34,
+			term:            6,
+			clusterVersion:  "3.7.0",
+			members:         []uint64{160, 200},
+			learners:        []uint64{300},
+		},
+		{
+			name:            "no learner and removed members",
+			consistentIndex: 19,
+			term:            5,
+			clusterVersion:  "3.6.0",
+			members:         []uint64{120, 220},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			dataDir := t.TempDir()
+			lg := zap.NewNop()
+
+			require.NoError(t, fileutil.TouchDirAll(lg, datadir.ToMemberDir(dataDir)))
+			require.NoError(t, fileutil.TouchDirAll(lg, datadir.ToWALDir(dataDir)))
+			require.NoError(t, fileutil.TouchDirAll(lg, datadir.ToSnapDir(dataDir)))
+
+			t.Log("Populate the wal file")
+			w, err := wal.Create(lg, datadir.ToWALDir(dataDir), pbutil.MustMarshal(
```

Comment:
Is it possible to have comment to describe what the initial state looks like?

Reply:
Updated.